### PR TITLE
appsec: SDK function for parsed http body instrumentation

### PR DESCRIPTION
**Pull request by Hellzy**
### Summary

This change allows users to monitor the `server.request.body` address by using an exported AppSec function.
The function, `MonitorParsedHTTPBody`, propagates to dyngo, starts a new operation and emits a start/finish event for this new operation type.

#### AppSec
- appsec/appsec.go: public appsec package used to export functions to users. `MonitorParsedHTTPBody` is there.
- appsec/example_test.go: examples showing how to use the new SDK for documentation purpose

- internal/appsec/dyngo/instrumentation/httpsec/http.go: add new `SDKBody` operation and core implementation for operation start/finish
- internal/appsec/waf.go: add SDKBody operation handling when event is caught. Fill `server.request.body` before calling the WAF
- internal/appsec/waf_test.go: add test case for rule using `server.request.body`

#### Contribs
- contrib/gin-gonic/gin/appsec.go: update code to reflect `StartOperation` API change and work around gin custom context
- contrib/labstack/echo.v4/appsec.go: update code to reflect `StartOperation` API change and work around echo custom context

#### Note
AppSec now updates the http request's context by inserting a key/val that allows keeping track of appsec operations

### CODE REVIEW

**Comment by Hellzy** [comment]
@Julio-Guerra I'd love your input, especially for:
- Naming of the top level appsec pkg
- Use of a map in http dyngo to track operations

File: appsec/example_test.go
Comment (Julio-Guerra) [inline appsec/example_test.go:33]: Following my previous comment:
```suggestion
		payload, err := myBodyParser(r.Body)
		if err != nil {
			http.Error(w, err.String(), http.StatusInternalServerError)
			return
		}
		appsec.MonitorParsedHTTPBody(r.Context(), payload)
		// ...
```

File: appsec/example_test.go
Comment (Julio-Guerra) [inline appsec/example_test.go:45]: Following my previous comment:
```suggestion
		req := c.Request()
		payload, err := myBodyParser(req.Body)
		if err != nil {
			http.Error(w, err.String(), http.StatusInternalServerError)
			return
		}
		appsec.MonitorParsedHTTPBody(req.Context(), payload)
		// ...
```

File: contrib/gin-gonic/gin/appsec.go
Comment (Julio-Guerra) [inline contrib/gin-gonic/gin/appsec.go:30]:
```suggestion
	ctx, op := httpsec.StartOperation(req.Context(), args)
```

File: contrib/gin-gonic/gin/appsec.go
Comment (Julio-Guerra) [inline contrib/gin-gonic/gin/appsec.go:31]: I think we avoid a struct copy by doing:
```suggestion
	c.Request = req.WithContext(ctx)
```

File: contrib/labstack/echo.v4/appsec.go
Comment (Julio-Guerra) [inline contrib/labstack/echo.v4/appsec.go:31]: Same as for gin:
```suggestion
		c.SetRequest(req.WithContext(ctx))
```

File: internal/appsec/dyngo/instrumentation/httpsec/http.go
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/http.go:55]:
```suggestion
	SDKBodyOperationRes struct {}
```

File: internal/appsec/dyngo/instrumentation/httpsec/http.go
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/http.go:50]:
```suggestion
		// Body corresponds to the address `server.request.body`.
		Body interface{}
```

File: appsec/appsec.go
Comment (Julio-Guerra) [inline appsec/appsec.go:24]: I doesn't compile, but as discussed privately, we should add:
```suggestion
	if !internal.Enabled() {
		// bonus: use sync.Once to log a debug message once, but I think it is fine not to.
		return
	}
	httpsec.MonitorParsedBody(ctx, body)
```
In order to better log in httpsec when the parent operation cannot be found, under the new assumption it should be called when appsec is disabled.

File: internal/appsec/dyngo/instrumentation/httpsec/http.go
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/http.go:58]:
```suggestion
// MonitorParsedBody starts and finishes the SDK body operation.
// This function should not be called when AppSec is disabled in order to
// get preciser error logs.
```

File: internal/appsec/dyngo/instrumentation/httpsec/http.go
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/http.go:65]: Under the new usage assumptions of calling this function only when appsec is enabled, we can now better log a legitimate error message:
```suggestion
		log.Error("appsec: parsed http body monitoring ignored: could not find the http handler instrumentation metadata in the request context: the request handler is not being monitored by a middleware function or the given context is not the expected request context")
```

File: internal/appsec/dyngo/instrumentation/httpsec/http.go
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/http.go:138]: This is unused and unecessary. It is in the HTTP operation in order to aggregate the list of events happening during the HTTP handler lifetime, including those that would happen in the body operation.
 
```suggestion
```

File: internal/appsec/dyngo/instrumentation/httpsec/http.go
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/http.go:156]:
```suggestion
	// Avoid a runtime panic in case of type-assertion error by collecting the 2 return values
```

File: internal/appsec/dyngo/instrumentation/httpsec/http.go
Comment (Julio-Guerra) [inline internal/appsec/dyngo/instrumentation/httpsec/http.go:145]:
```suggestion
// context and arguments and emits a start event up in the operation stack.
```

File: internal/appsec/waf.go
Comment (Julio-Guerra) [inline internal/appsec/waf.go:89]: You don't need to listen for the finish operation here:
```suggestion
			body = args.Body
```

**Review by Julio-Guerra — CHANGES_REQUESTED** [review]

File: appsec/example_test.go
Comment (Julio-Guerra) [inline appsec/example_test.go:27]:
```suggestion
```

**Review by Julio-Guerra — APPROVED** [review]
Approving with two more details suggested

**Review by Julio-Guerra — APPROVED** [review]

File: appsec/appsec.go
Comment (gbbr) [inline appsec/appsec.go:22]: If the context must be an HTTP request then why not just pass an HTTP request instead? This is quite error prone as it is now.

Lastly, what do you mean by a "parsed" HTTP request body? Can't we just avoid the `interface{}` type?

Comment (Hellzy) [reply appsec/appsec.go:22]: Thanks for looking into this.
> If the context must be an HTTP request then why not just pass an HTTP request instead?

I think I get your point, ensuring what we ask for by asking the user to pass the request directly rather than a context that could potentially be unrelated to the http request.
While this would work I'm not sure this is such a good way to enforce it, because:
- The request could be ill formed as well since it's provided by the user, meaning this is still error prone. The error vector is shifted but still there and it remains the responsibility of the implementation to handle errors properly.
- This becomes less future proof since some frameworks implement their own Request type (example: `contrib/gofiber/fiber.v2` that relies on `fasthttp`).

I'd be of the opinion to keep the prototype as is given that it allows more flexibility for future frameworks handling for AppSec, plus the fact that the user error possibility it introduces would still be there with a different prototype, only different. WDYT?


> Lastly, what do you mean by a "parsed" HTTP request body?

We indeed refer to the unmarshalled content of the request's body. We use `reflect` further down the stack to analyse the types and pass them to the WAF.

**Review by Julio-Guerra — DISMISSED** [review]

File: contrib/gin-gonic/gin/gintrace_test.go
Comment (knusbaum) [inline contrib/gin-gonic/gin/gintrace_test.go:18]: Nit: Let's keep package names lower case as is good go style.

File: contrib/gin-gonic/gin/gintrace_test.go
Comment (knusbaum) [inline contrib/gin-gonic/gin/gintrace_test.go:559]: Question: what is `"$globals"`?

**Review by knusbaum — CHANGES_REQUESTED** [review]
Minor nit and a question

File: contrib/gin-gonic/gin/gintrace_test.go
Comment (Hellzy) [reply contrib/gin-gonic/gin/gintrace_test.go:559]: `$globals` is a pattern that will trigger an attack detection in the WAF if found in the unmarshalled body.
It's a PHP variable that references all available variables in the scope. As an example, it could be used by an attacker to dump the server PHP environment: `echo "$globals"`.

Comment (knusbaum) [reply contrib/gin-gonic/gin/gintrace_test.go]: Thanks for the context. 👍 
It may be good to document that since it's relevant to the test, but I'll leave that up to you.

**Review by knusbaum — APPROVED** [review]

**Review by Julio-Guerra — APPROVED** [review]

**Pull request merged**
